### PR TITLE
Get instance global snapshot

### PR DESCRIPTION
**Pull request by wookieJ**

### CODE REVIEW

File: envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt [hunk 1]
```diff
@@ -71,6 +94,7 @@ class SnapshotDebugController(controlPlane: ControlPlane) {
     }
 
     class SnapshotNotFoundException : RuntimeException("snapshot missing")
+    class GlobalSnapshotNotFoundException : RuntimeException("Global snapshot missing")
```
Comment (MarcinFalkowski) [inline envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt]: Shouldn't we write a handler for this exception, returning 404? Just like `handleSnapshotMissing` below.

File: envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt [hunk 2]
```diff
@@ -51,6 +55,25 @@ class SnapshotDebugController(controlPlane: ControlPlane) {
         }
     }
 
+    @GetMapping("/snapshot/global")
+    fun globalSnapshot(@RequestParam xds: Boolean?): ResponseEntity<SnapshotDebugInfo> {
```
Comment (MarcinFalkowski) [inline envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt]: I think it should be tested, just like other endpoints in this class.

File: envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt [hunk 3]
```diff
@@ -51,6 +55,25 @@ class SnapshotDebugController(controlPlane: ControlPlane) {
...
+        val globalSnapshot = snapshotUpdater.getGlobalSnapshot()
+        return if (globalSnapshot?.adsSnapshot == null) {
+            throw GlobalSnapshotNotFoundException()
```
[the remark this answers is not in the document]
Comment (wookieJ) [reply envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt]: I've missed xds... Added

File: envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt (hunk 2 above)
Comment (wookieJ) [reply envoy-control-runner/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/debug/SnapshotDebugController.kt]: Tests added

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/SnapshotDebugTest.kt [hunk 4]
```diff
@@ -97,4 +97,43 @@ open class SnapshotDebugTest : EnvoyControlTestConfiguration() {
         // then
         assertThat(snapshot.found).isFalse()
     }
+
+    @Test
+    open fun `should return global snapshot debug info from xds`() {
+        // given
+        registerService(name = "echo")
```
Comment (MarcinFalkowski) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/SnapshotDebugTest.kt]: Is this needed for the test? If we register service "echo", we should use it and check if "echo" cluster and/or endpoint is in the snapshot. Or we should get rid of this line.

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/SnapshotDebugTest.kt [hunk 5]
```diff
@@ -97,4 +97,43 @@ open class SnapshotDebugTest : EnvoyControlTestConfiguration() {
...
+            assertThat(snapshot.snapshot["routes"]).isNotEmpty()
+            assertThat(snapshot.snapshot["endpoints"]).isNotEmpty()
+            assertThat(snapshot.snapshot["clusters"].first()["edsClusterConfig"]["edsConfig"].toString()).contains("envoy-control-xds")
+        }
+    }
+
+    @Test
+    open fun `should return global snapshot debug info from ads`() {
+        // given
+        registerService(name = "echo")
```
Comment (MarcinFalkowski) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/SnapshotDebugTest.kt]: as above

**Pull request merged**